File: src/commands/list.ts
/**
 * List command - List all servers and their tools
 */

import {
  type ToolInfo,
  type McpConnection,
  getConnection,
  debug,
  getConcurrencyLimit,
  safeClose,
} from '../client.js';
import {
  type McpServersConfig,
  getServerConfig,
  listServerNames,
  loadConfig,
} from '../config.js';
import { ErrorCode } from '../errors.js';
import { formatServerList } from '../output.js';

export interface ListOptions {
  withDescriptions: boolean;
  configPath?: string;
}

interface ServerWithTools {
  name: string;
  tools: ToolInfo[];
  instructions?: string;
  error?: string;
}

/**
 * Process items with limited concurrency, preserving order
 * Uses a worker pool pattern where each worker grabs the next item from a shared index
 */
async function processWithConcurrency<T, R>(
  items: T[],
  processor: (item: T, index: number) => Promise<R>,
  maxConcurrency: number,
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let currentIndex = 0;

  async function worker(): Promise<void> {
    while (currentIndex < items.length) {
      const index = currentIndex++;
      results[index] = await processor(items[index], index);
    }
  }

  // Start workers up to concurrency limit
  const workers = Array.from(
    { length: Math.min(maxConcurrency, items.length) },
    () => worker(),
  );

  await Promise.all(workers);
  return results;
}

/**
 * Fetch tools from a single server (uses daemon if enabled)
 */
async function fetchServerTools(
  serverName: string,
  config: McpServersConfig,
): Promise<ServerWithTools> {
  let connection: McpConnection | null = null;
  try {
    const serverConfig = getServerConfig(config, serverName);
    connection = await getConnection(serverName, serverConfig);

    const tools = await connection.listTools();
    const instructions = await connection.getInstructions();
    debug(`${serverName}: loaded ${tools.length} tools`);
    return { name: serverName, tools, instructions };
  } catch (error) {
    const errorMsg = (error as Error).message;
    debug(`${serverName}: connection failed - ${errorMsg}`);
    return {
      name: serverName,
      tools: [],
      error: errorMsg,
    };
  } finally {
    if (connection) {
      await safeClose(connection.close);
    }
  }
}

/**
 * Execute the list command
 */
export async function listCommand(options: ListOptions): Promise<void> {
  let config: McpServersConfig;

  try {
    config = await loadConfig(options.configPath);
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  const serverNames = listServerNames(config);

  if (serverNames.length === 0) {
    console.error(
      'Warning: No servers configured. Add servers to mcp_servers.json',
    );
    return;
  }

  const concurrencyLimit = getConcurrencyLimit();
  debug(
    `Processing ${serverNames.length} servers with concurrency ${concurrencyLimit}`,
  );

  // Process servers in parallel with concurrency limit
  const servers = await processWithConcurrency(
    serverNames,
    (name) => fetchServerTools(name, config),
    concurrencyLimit,
  );

  // Sort by name to ensure consistent output order
  servers.sort((a, b) => a.name.localeCompare(b.name));

  // Convert errors to tool-like display for human output
  const displayServers = servers.map((s) => ({
    name: s.name,
    instructions: s.instructions,
    tools: s.error
      ? [
          {
            name: `<error: ${s.error}>`,
            description: undefined,
            inputSchema: {},
          },
        ]
      : s.tools,
  }));

  // Human-readable output
  console.log(formatServerList(displayServers, options.withDescriptions));
}

File: src/commands/list.rs
use crate::config::debug;
use crate::errors::CliError;
use crate::output::ToolInfo;
use std::sync::Arc;
use tokio::sync::Semaphore;

pub struct ListOptions {
    pub with_descriptions: bool,
    pub config_path: Option<String>,
}

struct ServerWithTools {
    name: String,
    tools: Vec<ToolInfo>,
    instructions: Option<String>,
    error: Option<String>,
}

async fn fetch_server_tools(
    server_name: &str,
    config: &crate::config::McpServersConfig,
) -> ServerWithTools {
    let server_config = match crate::config::get_server_config(config, server_name) {
        Ok(sc) => sc,
        Err(e) => {
            return ServerWithTools {
                name: server_name.to_string(),
                tools: Vec::new(),
                instructions: None,
                error: Some(e.message),
            };
        }
    };

    match crate::client::get_connection(server_name, &server_config).await {
        Ok(connection) => {
            let tools_res = connection.list_tools().await;
            let inst_res = connection.get_instructions().await;

            let _ = connection.close().await;

            match tools_res {
                Ok(tools) => {
                    let instructions = inst_res.ok().flatten();
                    debug(&format!("{}: loaded {} tools", server_name, tools.len()));
                    ServerWithTools {
                        name: server_name.to_string(),
                        tools,
                        instructions,
                        error: None,
                    }
                }
                Err(e) => {
                    debug(&format!(
                        "{}: tools listing failed - {}",
                        server_name, e.message
                    ));
                    ServerWithTools {
                        name: server_name.to_string(),
                        tools: Vec::new(),
                        instructions: None,
                        error: Some(e.message),
                    }
                }
            }
        }
        Err(e) => {
            debug(&format!("{}: connection failed - {}", server_name, e.message));
            ServerWithTools {
                name: server_name.to_string(),
                tools: Vec::new(),
                instructions: None,
                error: Some(e.message),
            }
        }
    }
}

pub async fn list_command(options: ListOptions) -> Result<(), CliError> {
    let config = crate::config::load_config(options.config_path.as_deref())?;

    let server_names = crate::config::list_server_names(&config);

    if server_names.is_empty() {
        eprintln!("Warning: No servers configured. Add servers to mcp_servers.json");
        return Ok(());
    }

    let concurrency_limit = crate::config::get_concurrency_limit();
    debug(&format!(
        "Processing {} servers with concurrency {}",
        server_names.len(),
        concurrency_limit
    ));

    let semaphore = Arc::new(Semaphore::new(concurrency_limit));
    let mut tasks = Vec::new();

    for name in server_names {
        let name = name.clone();
        let config = config.clone();
        let sem = Arc::clone(&semaphore);

        tasks.push(tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            fetch_server_tools(&name, &config).await
        }));
    }

    let mut servers = Vec::new();
    for task in tasks {
        if let Ok(res) = task.await {
            servers.push(res);
        }
    }

    servers.sort_by(|a, b| a.name.cmp(&b.name));

    let display_servers: Vec<crate::output::ServerInfo> = servers
        .into_iter()
        .map(|s| {
            let tools = if let Some(err) = s.error {
                vec![ToolInfo {
                    name: format!("<error: {}>", err),
                    description: None,
                    input_schema: serde_json::json!({}),
                }]
            } else {
                s.tools
            };

            crate::output::ServerInfo {
                name: s.name,
                instructions: s.instructions,
                tools,
            }
        })
        .collect();

    println!(
        "{}",
        crate::output::format_server_list(&display_servers, options.with_descriptions)
    );
    Ok(())
}
